utility/mmsp2csv.cpp

```diff
@@ -0,0 +1,672 @@
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<unsigned long> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<unsigned long> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<unsigned long> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+			else if (long_type) {
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<long> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<long> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<long> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+			else if (unsigned_short_type) {
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<unsigned short> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<unsigned short> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<unsigned short> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+			else if (short_type) {
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<short> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<short> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<short> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+			else if (float_type) {
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<float> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<float> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<float> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+			else if (long_double_type) {
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<long double> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<long double> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<long double> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+			else if (double_type) {
+				if (dim == 1) {
+					MMSP::grid<1,MMSP::sparse<double> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 2) {
+					MMSP::grid<2,MMSP::sparse<double> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				} else if (dim == 3) {
+					MMSP::grid<3,MMSP::sparse<double> > GRID(argv[1]);
+					convert_sparses(GRID, csvfil);
+				}
+			}
+		}
+
+	return 0;
+}
+
+template <int dim, typename T> void convert_scalars(const MMSP::grid<dim,T>& GRID, std::ofstream& csvfil)
+{
+	if (dim==1) {
```

Comment:
These `dim==x` blocks are extraneous, since the contents are identical. Please remove them, simplifying to just one (*e.g.*, the body of the `if (dim == 1)` block), and test.